### PR TITLE
feat: Register custom metrics

### LINKED ISSUE #2061: feat(skore): Saving custom/extra metrics to reports
### Is your feature request related to a problem? Please describe.

Right now skore provides some default metrics available in a report with `report.metrics.<metric_name>` and also `report.metrics.summarize().frame()`. It'd be nice if a user could also save any extra metric they want their report to have and the report exposes access to. If I I have a `custom_metric`, in order to see it's results I need to do

```python
report.metrics.summarize(
    scoring=metric_function,
    scoring_kwargs={"response_method": "predict"},
)
```

 or 

```python
report.metrics.custom_metric(
    metric_function=metric_function,
    response_method="predict",
)
```

which requires me to have access to the `metric_function` function's definition. 

In addition, if I want to see `report.metrics.summarize().frame()`, I cannot see the default metrics + my metric by default. I'd need to pass all the metrics I want to `summarize(scoring=["rmse", "r2", metric_function])`. I think it would be nice if I could save a custom metric and whenever I do `report.metrics.summarize().frame()` all the metrics (the default ones by skore and mine) are shown. And also, if I share reports, I wouldn't necessarily expect the other user to have access to my `metric_function`, so if I could save my metric in some way into the report that'd be nice. 

### Describe the solution you'd like

If I want to save a custom metric to be part of my report, I'd like to be able to do something like

```python
report.metrics.save_custom_metric(name="my own metric", metric_function=metric_function)
```

Then I (or if I share the report, a colleague) can do:

```python
report.metrics.custom_metric("my_own_metric")
```

Also if I have a custom metric saved, I would expect to see it alongisde the provided by default metrics. (and maybe have a flag like `summarize(include_custom=True)`

```python
report.metrics.summarize().frame()
# RMSE ...
# R2 ...
# my_own_metric
```

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

from a comment on discord

### PR DESCRIPTION
**Pull request by auguste-probabl**
Depends on #2646, please review that one first

#### Change description

This PR makes it possible to register a custom metric in an `EstimatorReport`, so that it's shown along with the other default metrics in `summarize`:

```python
from sklearn.datasets import load_breast_cancer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import make_scorer, mean_absolute_error
from skore import evaluate
X, y = load_breast_cancer(return_X_y=True)
classifier = LogisticRegression(max_iter=10_000)
report = evaluate(classifier, X, y, splitter=0.2, pos_label=1)
report.metrics.register(
    make_scorer(mean_absolute_error, response_method="predict")
)
report.metrics.summarize().frame()

                     LogisticRegression
Metric
Accuracy                       0.947368
Precision                      0.984127
Recall                         0.925373
ROC AUC                        0.993649
Brier score                    0.036154
Fit time (s)                   0.324200
Predict time (s)               0.000323
Mean Absolute Error            0.052632
```

Closes #2061

#### Contribution checklist

- [x] Unit tests were added or updated (if necessary)
- [ ] Documentation was added or updated (if necessary)
- [ ] The documentation builds and renders properly (if it does, our bot will add a comment linking
to a preview of the documentation to review it visually)
- [ ] A new changelog entry was added to CHANGELOG.rst (if necessary; typically, if your change
requires updating tests)

#### AI usage disclosure

AI tools were involved for:
- [x] Code generation (e.g., when writing an implementation or fixing a bug)
- [x] Test/benchmark generation
- [ ] Documentation (including examples)
- [x] Research and understanding

I had an LLM write down specs in docs/design, write some tests based on those specs, and some of the implementation.

### CODE REVIEW

File: docs/design/0007-introduce-metrics-registry.md
Comment (auguste-probabl) [inline docs/design/0007-introduce-metrics-registry.md:1]: maybe delete this, or spend a lot more time syncing them with the actual implementation

File: skore/tests/unit/reports/estimator/metrics/test_summarize_new.py
Comment (auguste-probabl) [inline skore/tests/unit/reports/estimator/metrics/test_summarize_new.py:1]: merge this into test_summarize, or replace it entirely

Comment (auguste-probabl) [inline skore/tests/unit/reports/estimator/metrics/test_summarize_new.py:1]: ensure tests follow our conventions

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py [hunk 1]
```diff
@@ -24,41 +22,280 @@
...
+        """
+        parsed = self._parse_metric(metric, {})
+
+        # Protect built-in names
+        if parsed.name in self._registry and self._registry[parsed.name].is_builtin:
+            raise ValueError(
+                f"Cannot register {parsed.name!r}: it is a built-in metric name."
+            )
+
+        # Warn for lambdas and closures (not pickle-safe).
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_estimator/metrics_accessor.py:141]: ensure the source code-grabbing functionality works as intended

Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_estimator/metrics_accessor.py:141]: ensure this logic properly fixes #2060 (check if a non-regression test is there already)

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_estimator/metrics_accessor.py:1]: changelog

Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_estimator/metrics_accessor.py:1]: docs (api docs, a short example)

Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_estimator/metrics_accessor.py:645]: currently `summarize` calls the individual metric methods, e.g. `accuracy()`. I wonder if we could reverse that dependency, and make `accuracy()` call `summarize` instead.

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py [hunk 2]
```diff
@@ -24,41 +22,280 @@
     PredictionErrorDisplay,
     RocCurveDisplay,
 )
+from skore._sklearn.metrics import (
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_estimator/metrics_accessor.py:28]: test interaction with Project

Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_estimator/metrics_accessor.py:28]: test integration with skore-hub-project

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py
Comment (auguste-probabl) [reply skore/src/skore/_sklearn/_estimator/metrics_accessor.py:645]: That might fix #2001

Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_estimator/metrics_accessor.py:1]: maybe fix #2203 in this PR as well

Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_estimator/metrics_accessor.py:1]: Need to split this into (1) Refactor to registry pattern (2) add `register`

**Comment by jeromedockes** [comment]
is this superseded by #2646 ?

**Comment by auguste-probabl** [comment]
> is this superseded by #2646 ?

Sorry for the misunderstanding; this PR *depends* on #2646; it contains all of its commits, and will become much smaller when #2646 is merged.
